bdns/dns.go

```diff
@@ -475,3 +503,60 @@ func (dnsClient *DNSClientImpl) LookupCAA(ctx context.Context, hostname string)
 	}
 	return CAAs, nil
 }
+
+// logDNSError logs the provided err result from making a query for hostname to
+// the chosenServer. If the err is a `dns.ErrId` instance then the raw Base64
+// URL encoded bytes of the query (and if not-nil, the response) in wire format
+// is logged as well. This function is called from exchangeOne only for the case
+// where an error occurs querying a hostname that indicates a problem between
+// the VA and the chosenServer.
+func logDNSError(
+	logger blog.Logger,
+	chosenServer string,
+	hostname string,
+	msg, resp *dns.Msg,
+	underlying error) {
+	// We don't expect logDNSError to be called with a nil msg or err but
+	// if it happens return early. We allow resp to be nil.
+	if msg == nil || len(msg.Question) == 0 || underlying == nil {
+		return
+	}
+	queryType := dns.TypeToString[msg.Question[0].Qtype]
+
+	// If the error indicates there was a query/response ID mismatch then we want
+	// to log more detail.
+	if underlying == dns.ErrId {
+		packedMsgBytes, err := msg.Pack()
+		if err != nil {
+			logger.Errf("logDNSError failed to pack msg: %v\n", err)
+			return
+		}
+		encodedMsg := base64.RawURLEncoding.EncodeToString(packedMsgBytes)
```

Comment:
For this type of use I have a preference for `StdEncoding` since there are easy commandline tools to decode it, and we don't need to stick things in URLs.

Reply:
:+1: Changed to `StdEncoding` in 44cd682